### ISSUE TITLE
Indexing when len(A_train) > len(B_train)?

### ISSUE THREAD

**Issue by kpagels** [junyanz/pytorch-CycleGAN-and-pix2pix#818]
I tried to look into your code but wasnt able to find exactly where this is located. 

How do you guys index when len(A_train) > len(B_train) or the other way around? Lets say len(A_train) = 1000, len(B_train) = 2000 and batchsize = 16. I assume you index like A_train(1:16), A_train(17:33) etc. What happens when A_train( 976:993) passed? Will you choose random indexing?

And can you maybe point me to the location of the code? 

Thanks!

**Comment by junyanz** [comment]
Here is the [code](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/data/unaligned_dataset.py#L51) for random indexing.  Here is the [code](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/data/unaligned_dataset.py#L65) to get the length of the dataset.  We have two ways of indexing, which depends on the flag `--serial_batches`.

**Comment by kpagels** [comment]
Thanks for your answer @junyanz. I have some followup questions based on that. And I hope you have time to answer me.

1) It seems like the code you are referring to only is used when the batch_size = 1? I base this on index_B is only one integer value? Where do you use the class UnalignedDataset (I wasn't able to locate it)? Is this called every time training is done? I'm curious because I would love to see what is done when batch_size > 1. 

1) I assume the **_index_** is a random integer for data indexing. So that means by default that the indexing start point is random, right? And if this is true, I don't see any difference between the following two lines:
The following index_B is random because **_index_** is random.:
`index_B = index % self.B_size`
The following indexing is random because of randint:
`index_B = random.randint(0, self.B_size - 1)`
Won't it just be random index by default anyway because of variable **_index_**?

3) So you are using the flag **_serial_batches_** and its _**False**_ by default. As I understand it this means that the indexing when training is random e.g. 1,6,100,2,600 etc. which correspond to shuffling the data? And if _**serial_batches**_ = True the order is 1,2,3,4,5 etc.? Right?

4) If the above is true: don't you shuffle the data before its used for training? Or is that what the flag serial_batches determines? By calling the function __get_item__ [code](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/data/unaligned_dataset.py#L39) and determine if the dataset should be shuffled or not?

Thanks!

**Comment by junyanz** [comment]
1. It works `batch_size > 1`. The Pytorch built-in data loader will automatically handle it. 
2. The difference is whether index_B depends on index_A or not. In the first line, it depends. In the second line, it doesn't. 
3.  Nope. See 2. The index is a random integer assigned by the PyTorch data loader. 
4. PyTorch dataloader will shuffle the data for you (if setting `shuffle=True`). You don't need to shuffle it by yourself.

For more details, please refer to the [document](https://pytorch.org/docs/stable/data.html) for the PyTorch data loader.

**Comment by kpagels** [comment]
Thanks!
